superset-frontend/src/views/CRUD/data/database/DatabaseModal.tsx

```diff
@@ -82,53 +116,147 @@ const DatabaseModal: FunctionComponent<DatabaseModalProps> = ({
+  } = useSingleViewResource<DatabaseObject>(
+    'database',
+    t('database'),
+    addDangerToast,
+  );
+
   // Functions
   const hide = () => {
     setIsHidden(true);
     onHide();
   };
 
   const onSave = () => {
-    if (onDatabaseAdd) {
-      onDatabaseAdd();
-    }
+    if (isEditMode) {
+      // Edit
+      const update: DatabaseObject = {
+        database_name: db ? db.database_name : '',
+        sqlalchemy_uri: db ? db.sqlalchemy_uri : '',
+        ...db,
+      };
+
+      // Need to clean update object
+      if (update.id) {
+        delete update.id;
+      }
+
+      if (!update.cache_timeout) {
+        update.cache_timeout = '0';
+      }
 
-    hide();
+      if (!update.encrypted_extra) {
+        update.encrypted_extra = '';
+      }
+
+      if (!update.force_ctas_schema) {
+        update.force_ctas_schema = '';
+      }
+
+      if (!update.server_cert) {
+        update.server_cert = '';
+      }
+
+      if (db && db.id) {
+        updateResource(db.id, update).then(() => {
+          if (onDatabaseAdd) {
+            onDatabaseAdd();
+          }
+
+          hide();
+        });
+      }
+    } else if (db) {
+      // Create
+      createResource(db).then(() => {
+        if (onDatabaseAdd) {
+          onDatabaseAdd();
+        }
+
+        hide();
+      });
+    }
   };
 
   const onInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
     const target = event.target;
     const data = {
       database_name: db ? db.database_name : '',
-      uri: db ? db.uri : '',
+      sqlalchemy_uri: db ? db.sqlalchemy_uri : '',
       ...db,
     };
 
-    data[target.name] = target.value;
+    if (target.type === 'checkbox') {
+      data[target.name] = target.checked;
+    } else {
+      data[target.name] = target.value;
+    }
 
     setDB(data);
   };
 
-  const isEditMode = database !== null;
+  const onTextChange = (event: React.ChangeEvent<HTMLTextAreaElement>) => {
+    const target = event.target;
+    const data = {
+      database_name: db ? db.database_name : '',
+      sqlalchemy_uri: db ? db.sqlalchemy_uri : '',
+      ...db,
+    };
+
+    data[target.name] = target.value;
+    setDB(data);
+  };
+
+  const validate = () => {
+    if (
+      db &&
+      db.database_name.length &&
+      db.sqlalchemy_uri &&
+      db.sqlalchemy_uri.length
+    ) {
+      setDisableSave(false);
+    } else {
+      setDisableSave(true);
+    }
+  };
 
   // Initialize
   if (
     isEditMode &&
     (!db || !db.id || (database && database.id !== db.id) || (isHidden && show))
   ) {
-    setDB(database);
+    if (database && database.id !== null && !dbLoading) {
+      const id = database.id || 0;
+
+      fetchResource(id).then(() => {
+        setDB(dbFetched);
+      });
+    }
   } else if (!isEditMode && (!db || db.id || (isHidden && show))) {
     setDB({
       database_name: '',
-      uri: '',
+      sqlalchemy_uri: '',
     });
   }
 
+  // Validation
+  useEffect(() => {
+    validate();
```

Comment:
Do we really need to call this on every render? It's no an expensive operation now, but it could become one. Would adding a dependency on `db` be a sensible optimization?